### ISSUE TITLE
New feature request: LOBPCG in addition to Lanczos?

### ISSUE THREAD

**Issue by lobpcg** [tensorflow/tensorflow#21835]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
N/A
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
all
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
all
- **TensorFlow installed from (source or binary)**:
N/A
- **TensorFlow version (use command below)**:
N/A
- **Python version**:
N/A
- **Bazel version (if compiling from source)**:
N/A
- **GCC/Compiler version (if compiling from source)**:
N/A
- **CUDA/cuDNN version**:
N/A
- **GPU model and memory**:
N/A
- **Exact command to reproduce**:
N/A

### Describe the problem
TensorFlow now has a native implementation of Lanczos https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/solvers/python/ops/lanczos.py

It could be useful to add a native TensorFlow implementation of LOBPCG, see https://en.wikipedia.org/wiki/LOBPCG which is an alternative to Lanczos  and has some advantages, e.g., warm-starts. SciPy has the Python native implementation https://docs.scipy.org/doc/scipy-1.1.0/reference/generated/scipy.sparse.linalg.lobpcg.html used in Scikit for manifold spectral embedding http://scikit-learn.org/stable/modules/generated/sklearn.manifold.spectral_embedding.html and for spectral clustering http://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html

Reference implementations of LOBPCG are described in 

1. Knyazev, Andrew V. (2001). "Toward the Optimal Preconditioned Eigensolver: Locally Optimal Block Preconditioned Conjugate Gradient Method". SIAM Journal on Scientific Computing. 23 (2): 517–541. doi:10.1137/S1064827500366124
2. Knyazev, A. V.; Argentati, M. E.; Lashuk, I.; Ovtchinnikov, E. E. (2007). "Block Locally Optimal Preconditioned Eigenvalue Xolvers (BLOPEX) in Hypre and PETSc". SIAM Journal on Scientific Computing. 29 (5): 2224. arXiv:0705.2626 Freely accessible. doi:10.1137/060661624

LOBPCG can be easily adopted to compute partial SVD and PCA for a data matrix X without ever computing its covariance matrix X'*X, i.e. in matrix-free fashion, see comments at https://www.mathworks.com/matlabcentral/fileexchange/48-lobpcg-m

NVIDIA has implemented LOBPCG in its nvGRAPH library introduced in CUDA 8.

A first simple step could be to write a single-vector (non-block) version LOPCG, which would be very similar to already existing https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/solvers/python/ops/linear_equations.py
but instead of solving Ax=rhs would produce a single eigenvector of selfadjoint matrix `A`  matrix-free where the action of the matrix A is represented by  `operator`, corresponding to the largest (or smallest) eigenvalue, possibly constrained to be orthogonal to a set of given vectors Y. The latter would allow computing several main eigenvectors one-by-one, putting all previously computed eigenvectors into Y.

I can help with the implementation if numerical issues arise.

### Source code / logs
N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by lobpcg** [comment]
Done adding N/A

**Comment by karmel** [comment]
@rmlarsen , can you comment on the feasibility of this feature request?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by lobpcg** [comment]
We'll try to do it in a good way in https://github.com/lobpcg/blopex-python/ and get back here when done.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
